[export] specialize for aten.to

pianpwk opened the pull request:
Changes decomposition behavior of `aten.to` to respect the aliasing/non-aliasing behavior in eager, and to specialize to the input/conversion dtype & device.

Before change: we always decompose `aten.to` into `_to_copy`, regardless of aliasing behavior. This leads us to ban mutations on the result of `_to_copy` when aliased, since we can't guarantee correct program semantics. This meant users had to explicitly call `.clone()` before mutating. In the special cases where we don’t ban mutations (e.g. dtype conversion), we add runtime assertions on the input & conversion dtype/devices in the decomposed program (see https://github.com/pytorch/pytorch/pull/142420).

After change: we decompose to the aliasing/non-aliasing behavior that matches eager, allowing mutations in all cases. We also add dtype/device assertions for all `aten.to` ops, starting in the pre-dispatch graph, basically specializing the program to the dtype/devices.

Differential Revision: D71229547

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

tugsbayasgalan commented on torch/export/exported_program.py:
```diff
@@ -278,21 +276,6 @@ def _force_dispatch_to_orig_cia_callable(fake_tensor_mode, op, *args, **kwargs):
-            torch.ops.aten.to.dtype_layout: _special_op_to_preserve_cia,
-            torch.ops.aten.to.dtype: _special_op_to_preserve_cia,
-        },
-        safe=False,
```
Can you get rid of safe flag from _override_composite_implicit_decomp? This was the only place it mattered.

yushangdi commented on test/distributed/tensor/test_dtensor_compile.py:
```diff
@@ -157,6 +157,7 @@ def forward(self, x):
             str(ep.graph_module.code).strip(),
             """\
 def forward(self, b_buffer, x):
+    _assert_tensor_metadata_default = torch.ops.aten._assert_tensor_metadata.default(x, dtype = torch.float64, device = device(type='cpu'), layout = torch.strided);  _assert_tensor_metadata_default = None
```
not requesting a change on the PR, just for my own understanding.

If we are checking the device here, then we're specializing to a specific device. So if the ep is exported with example tensors on cpu, the can only be used for cpu tensors in the future. Is this correct?

I thought we're making the export result device agnostic, just to check my understanding here. Thanks!

pianpwk replied:
Yeahh, this is quite the opposite of dtype/device agnostic export, we're specializing on both now.

pianpwk committed:
[export] specialize for aten.to (#149235)

Summary:
Changes decomposition behavior of aten.to to respect the aliasing/non-aliasing behavior in eager, and to specialize to the input/conversion dtype & device.

Before change: we always decompose `aten.to` into `_to_copy`, regardless of aliasing behavior. This leads us to ban mutations on the result of `_to_copy` when aliased, since we can't guarantee correct program semantics. This meant users had to explicitly call `.clone()` before mutating. In the special cases where we don’t ban mutations (e.g. dtype conversion), we add runtime assertions on the input & conversion dtype/devices in the decomposed program (see https://github.com/pytorch/pytorch/pull/142420).

After change: we decompose to the aliasing/non-aliasing behavior that matches eager, allowing mutations in all cases. We also add dtype/device assertions for all `aten.to` ops, starting in the pre-dispatch graph, basically specializing the program to the dtype/devices.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/149235

Test Plan: test_export

Reviewed By: tugsbayasgalan, shruthign

Differential Revision: D71229547
